app/app.go

```diff
@@ -1169,43 +1182,61 @@ func (d *Deployments) AbortDeployment(ctx context.Context, deploymentID string)
 
 func (d *Deployments) DecommissionDevice(ctx context.Context, deviceId string) error {
 
-	if err := d.db.DecommissionDeviceDeployments(ctx,
-		deviceId); err != nil {
-
-		return err
-	}
-
-	//get all affected deployments and update its stats
-	deviceDeployments, err := d.db.FindAllDeploymentsForDeviceIDWithStatuses(
+	var lastDeployment *time.Time
+	// Retrieve active device deployment for the device
+	deviceDeployment, err := d.db.FindOldestDeploymentForDeviceIDWithStatuses(
 		ctx,
-		deviceId, model.DeviceDeploymentStatusDecommissioned)
-
+		deviceId,
+		model.ActiveDeploymentStatuses()...)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "Searching for active deployment for the device")
+	} else if deviceDeployment != nil {
+		now := time.Now()
+		ddStatus := model.DeviceDeploymentStatus{
+			Status:     model.DeviceDeploymentStatusDecommissioned,
+			FinishTime: &now,
+		}
+		if err := d.UpdateDeviceDeploymentStatus(ctx, *deviceDeployment.DeploymentId,
+			deviceId, ddStatus); err != nil {
+			return errors.Wrap(err, "updating device deployment status")
+		}
+		lastDeployment = deviceDeployment.Created
+	} else {
+		//get latest device deployment for the device;
+		deviceDeployment, err := d.db.FindLatestDeploymentForDeviceIDWithStatuses(
+			ctx,
+			deviceId,
+			model.InactiveDeploymentStatuses()...)
+		if err != nil {
+			return errors.Wrap(err, "Searching for latest active deployment for the device")
+		} else if deviceDeployment == nil {
+			lastDeployment = &time.Time{}
+		} else {
+			lastDeployment = deviceDeployment.Created
+		}
 	}
 
-	for _, deviceDeployment := range deviceDeployments {
-
-		stats, err := d.db.AggregateDeviceDeploymentByStatus(
-			ctx, *deviceDeployment.DeploymentId)
+	// get deployments newer then last device deployment
+	// iterate over deployments and check if the device is part of the deployment or not
+	// if the device is part of the deployment create new, decommisioned device deployment
+	for skip := 0; true; skip++ {
+		deployments, err := d.db.FindNewerActiveDeployments(ctx, lastDeployment, skip, 1)
```

Comment:
sorry for the silly question: is it really needed to run N queries, where N is the number of active deployments? wouldn't it work the same way using pagination with, let's say, 100 records per page? Not a big deal, but it's a (micro?) optimization if you have several tens (or hundreds) of active deployments.

Reply:
yup; we can set a limit to 100 if that's the best number;
in the past, we were expecting tenants to have several, not several tens or hundreds, of active deployments, but it looks like it's not the case (after looking into hM data);
it would be good if everyone will take a look at the real hM data to have a general overview of how customers are using mender IMO

Reply:
done

Reply:
@kjaskiewiczz we have a few tenants with tens of active deployments (just verified), but we expect it to grow when they'll start using dynamic deployments, so better to reduce the number of queries. Thanks for addressing it!